lib/std/crypto/25519/field.zig

```diff
@@ -0,0 +1,327 @@
+    }
+
+    pub inline fn edwards25519d2() Fe {
+        return .{ .limbs = .{ 1859910466990425, 932731440258426, 1072319116312658, 1815898335770999, 633789495995903 } }; // 2d
+    }
+
+    // 1/sqrt(a-d)
+    pub inline fn edwards25519sqrtamd() Fe {
+        return .{ .limbs = .{ 278908739862762, 821645201101625, 8113234426968, 1777959178193151, 2118520810568447 } };
+    }
+
+    pub inline fn isZero(fe: Fe) bool {
+        var reduced = fe;
+        reduced.reduce();
+        const limbs = reduced.limbs;
+        return (limbs[0] | limbs[1] | limbs[2] | limbs[3] | limbs[4]) == 0;
+    }
+
+    pub inline fn equivalent(a: Fe, b: Fe) bool {
+        return a.sub(b).isZero();
+    }
+
+    pub fn fromBytes(s: [32]u8) Fe {
+        var fe: Fe = undefined;
+        fe.limbs[0] = readIntLittle(u64, s[0..8]) & MASK51;
+        fe.limbs[1] = (readIntLittle(u64, s[6..14]) >> 3) & MASK51;
+        fe.limbs[2] = (readIntLittle(u64, s[12..20]) >> 6) & MASK51;
+        fe.limbs[3] = (readIntLittle(u64, s[19..27]) >> 1) & MASK51;
+        fe.limbs[4] = (readIntLittle(u64, s[24..32]) >> 12) & MASK51;
+
+        return fe;
+    }
+
+    pub fn toBytes(fe: Fe) [32]u8 {
+        var reduced = fe;
+        reduced.reduce();
+        var s: [32]u8 = undefined;
+        writeIntLittle(u64, s[0..8], reduced.limbs[0] | (reduced.limbs[1] << 51));
+        writeIntLittle(u64, s[8..16], (reduced.limbs[1] >> 13) | (reduced.limbs[2] << 38));
+        writeIntLittle(u64, s[16..24], (reduced.limbs[2] >> 26) | (reduced.limbs[3] << 25));
+        writeIntLittle(u64, s[24..32], (reduced.limbs[3] >> 39) | (reduced.limbs[4] << 12));
+
+        return s;
+    }
+
+    pub fn rejectNonCanonical(s: [32]u8, comptime ignore_extra_bit: bool) !void {
+        var c: u16 = (s[31] & 0x7f) ^ 0x7f;
+        comptime var i = 30;
+        inline while (i > 0) : (i -= 1) {
+            c |= s[i] ^ 0xff;
+        }
+        c = (c -% 1) >> 8;
+        const d = (@intCast(u16, 0xed - 1) -% @intCast(u16, s[0])) >> 8;
+        const x = if (ignore_extra_bit) 0 else s[31] >> 7;
+        if ((((c & d) | x) & 1) != 0) {
+            return error.NonCanonical;
+        }
+    }
+
+    fn reduce(fe: *Fe) void {
+        comptime var i = 0;
+        comptime var j = 0;
+        const limbs = &fe.limbs;
+        inline while (j < 2) : (j += 1) {
+            i = 0;
+            inline while (i < 4) : (i += 1) {
+                limbs[i + 1] += limbs[i] >> 51;
+                limbs[i] &= MASK51;
+            }
+            limbs[0] += 19 * (limbs[4] >> 51);
+            limbs[4] &= MASK51;
+        }
+        limbs[0] += 19;
+        i = 0;
+        inline while (i < 4) : (i += 1) {
+            limbs[i + 1] += limbs[i] >> 51;
+            limbs[i] &= MASK51;
+        }
+        limbs[0] += 19 * (limbs[4] >> 51);
+        limbs[4] &= MASK51;
+
+        limbs[0] += 0x8000000000000 - 19;
```

Comment:
Could you confirm that this isn't meant to be wrapping arithmetic?

Reply:
Yes, it is critical that limbs never overflow. Wrapping would mean that the carry would get silently lost.